domain/application/state/state.go

```diff
@@ -96,67 +97,74 @@ AND revision = $charmReferenceNameRevisionSource.revision
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmMetadata(ctx, tx, uuid, charm.Metadata, charm.LXDProfile); err != nil {
+	if err := s.setCharmMetadata(ctx, tx, uuid, ch.Metadata, ch.LXDProfile); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmTags(ctx, tx, uuid, charm.Metadata.Tags); err != nil {
+	if err := s.setCharmTags(ctx, tx, uuid, ch.Metadata.Tags); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmCategories(ctx, tx, uuid, charm.Metadata.Categories); err != nil {
+	if err := s.setCharmCategories(ctx, tx, uuid, ch.Metadata.Categories); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmTerms(ctx, tx, uuid, charm.Metadata.Terms); err != nil {
+	if err := s.setCharmTerms(ctx, tx, uuid, ch.Metadata.Terms); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmRelations(ctx, tx, uuid, charm.Metadata); err != nil {
+	if err := s.setCharmRelations(ctx, tx, uuid, ch.Metadata); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmExtraBindings(ctx, tx, uuid, charm.Metadata.ExtraBindings); err != nil {
+	if err := s.setCharmExtraBindings(ctx, tx, uuid, ch.Metadata.ExtraBindings); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmStorage(ctx, tx, uuid, charm.Metadata.Storage); err != nil {
+	if err := s.setCharmStorage(ctx, tx, uuid, ch.Metadata.Storage); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmDevices(ctx, tx, uuid, charm.Metadata.Devices); err != nil {
+	if err := s.setCharmDevices(ctx, tx, uuid, ch.Metadata.Devices); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmPayloads(ctx, tx, uuid, charm.Metadata.PayloadClasses); err != nil {
+	if err := s.setCharmPayloads(ctx, tx, uuid, ch.Metadata.PayloadClasses); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmResources(ctx, tx, uuid, charm.Metadata.Resources); err != nil {
+	if err := s.setCharmResources(ctx, tx, uuid, ch.Metadata.Resources); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmContainers(ctx, tx, uuid, charm.Metadata.Containers); err != nil {
+	if err := s.setCharmContainers(ctx, tx, uuid, ch.Metadata.Containers); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmActions(ctx, tx, uuid, charm.Actions); err != nil {
+	if err := s.setCharmActions(ctx, tx, uuid, ch.Actions); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmConfig(ctx, tx, uuid, charm.Config); err != nil {
+	if err := s.setCharmConfig(ctx, tx, uuid, ch.Config); err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := s.setCharmManifest(ctx, tx, uuid, charm.Manifest); err != nil {
+	if err := s.setCharmManifest(ctx, tx, uuid, ch.Manifest); err != nil {
 		return errors.Trace(err)
 	}
 
+	// Insert the download info if the charm is from CharmHub.
+	if ch.Source == charm.CharmHubSource {
+		if err := s.setCharmDownloadInfo(ctx, tx, uuid, downloadInfo); err != nil {
```

Comment:
There's something I'm missing here. Why are we setting the download info on `SetCharm()`? Shouldn't this be done earlier in the lifecycle of a deploy?

Reply:
With bundles, we call `SetCharm` for uploading a charm and then `CreateApplication` is then associated with an existing charm.